File: cli/console.go
package cli

import (
	"context"
	"errors"
	"io"
	"os"
	"strings"
	"sync"

	"charm.land/bubbles/v2/textinput"
	tea "charm.land/bubbletea/v2"
	"charm.land/lipgloss/v2"
	"github.com/charmbracelet/colorprofile"
	"github.com/charmbracelet/x/term"
	"github.com/missingstudio/eva/cli/render"
	"github.com/missingstudio/eva/core"
	"github.com/missingstudio/eva/events"
)

// console is the interactive interface: what a person is typing, what is
// arriving, and the turns already answered.
//
// It consumes and it does not drive. The core loop runs the turn and the
// Recorder publishes what it committed; those Events reach this model from
// outside its own update loop, sent in by the feed below. That is what keeps
// the interface a consumer rather than the architecture — it holds no
// Provider, no Recorder, and no way to reach the Trace, so it cannot show a
// turn the record does not hold.
type console struct {
	// ask runs one turn and returns when the Run is closed. It is a function
	// rather than the assembly behind it, because everything the interface
	// could do with a Provider or a sink is something it must not do.
	ask func(ctx context.Context, prompt string) (core.Outcome, error)

	// control is what a command may change about the turns that follow.
	control control

	// renderer is the projection a person reads: a fold over committed Events,
	// showing on this model rather than writing to a stream.
	renderer *render.Renderer

	// ctx is what a Run is started under. The one a turn actually runs under
	// is a cancellable child of it, which is what Ctrl-C ends.
	ctx context.Context

	input textinput.Model

	// arriving is the open Run's answer as it streams, and it is what the
	// live area shows. The rendered turn that replaces it comes from the
	// Renderer, once the Run has closed — a heading, a list, and a fenced
	// block are each only meaningful once they are complete.
	arriving strings.Builder

	// pending holds what is waiting to go above the view. The program puts it
	// there, and only the update loop may ask it to.
	pending []string

	// cancel ends the Run in flight, and is nil when there is none.
	cancel context.CancelFunc

	// running counts the Runs whose goroutine has not returned. The process
	// waits on it before it closes the sink, because a Run that is still
	// committing when its Trace closes is the one failure the project has no
	// instrument to detect.
	running sync.WaitGroup

	// profile is what the destination can show. Colour is reduced to it here,
	// before a turn is handed over, for the reason put sets out.
	profile colorprofile.Profile

	// height is the window's, and is zero until the program reports one.
	height int

	// dark is what the terminal answered when it was asked what colour it is.
	// It starts at the assumption and is corrected by the answer.
	dark   bool
	styles styles
}

var _ tea.Model = (*console)(nil)

// styles are the interface's own, which are not the answer's. The answer is
// styled by the Renderer, and everything here is subordinate to it: a prompt
// echoed back and a status line are landmarks a person navigates by, not
// things they read.
type styles struct {
	said lipgloss.Style
	hint lipgloss.Style
}

func newStyles(dark bool) styles {
	return styles{
		said: lipgloss.NewStyle().Bold(true),
		// The same grey the cost line is written in. Both are subordinate to
		// the answer, so both are one decision rather than two that happen to
		// agree — see render.Subdued.
		hint: render.Subdued(dark),
	}
}

// committed carries one committed Event into the update loop from outside it.
type committed struct{ event events.Event }

// chunk carries one piece of an answer into the update loop ahead of the
// record of it. It is what the live area is made of, and it is never what a
// turn leaves behind — see Turn.Arriving.
type chunk struct{ text string }

// answered says a Run has closed, whatever it claimed.
type answered struct {
	outcome core.Outcome
	err     error
}

// feed carries committed Events to the interface.
//
// It is a Subscriber, so what it carries is what the Trace holds rather than
// what a producer sent, and it carries it after the commit rather than before.
// The program's own doorway for messages from other goroutines is what it
// sends through, which is the whole of the coupling between the core loop and
// the interface: one direction, one message type, and nothing shared.
type feed struct{ program *tea.Program }

var _ core.Subscriber = (*feed)(nil)

// Committed sends one Event to the interface.
//
// It cannot fail. A program that has ended takes no more messages and says so
// by returning, and a turn that is still committing to the Trace is doing the
// thing that matters whether or not anyone is watching.
func (f *feed) Committed(_ context.Context, e events.Event) error {
	f.program.Send(committed{event: e})
	return nil
}

// ended is an input stream, plus the end of it.
//
// A console reads keys until there are none left. Nothing in the program says
// an input stream has ended — it is a reader, and a reader that returns EOF
// simply stops being read — so `eva` with its input closed would sit at a
// prompt nobody can type at. This is what closes it, and it is why `eva` in a
// script exits rather than hangs.
type ended struct {
	// The whole file rather than a reader over it. A program takes the
	// terminal into raw mode by its descriptor, and it looks for the
	// descriptor on what it was handed — so a wrapper that hid one would
	// leave the terminal in cooked mode, where a line only arrives on enter,
	// where the terminal echoes it, and where Ctrl-C is a signal that kills
	// the process rather than a key that ends a turn.
	*os.File

	quit func()
}

func (e ended) Read(p []byte) (int, error) {
	n, err := e.File.Read(p)
	if errors.Is(err, io.EOF) {
		e.quit()
	}
	return n, err //nolint:wrapcheck // a reader's error is the reader's.
}

// endedFile is ended's counterpart for a descriptor [pollable] says cannot be
// handed over: the same "quit on EOF" behaviour, but as a bare reader, so the
// program never learns this input had a descriptor to begin with.
type endedFile struct {
	io.Reader
	quit func()
}

func (e endedFile) Read(p []byte) (int, error) {
	n, err := e.Reader.Read(p)
	if errors.Is(err, io.EOF) {
		e.quit()
	}
	return n, err //nolint:wrapcheck // a reader's error is the reader's.
}

// pollable is whether the operating system can watch this descriptor for
// readiness rather than fabricate it: a terminal or a pipe say so honestly, but
// a plain file and the null device report readiness without meaning it, and
// asking the OS to watch one for it fails outright rather than degrading. That
// failure is the program's to avoid, not the input's to surface, so this is
// checked before the descriptor is ever handed over.
func pollable(file *os.File) bool {
	if term.IsTerminal(file.Fd()) {
		return true
	}
	info, err := file.Stat()
	if err != nil {
		return false
	}
	return info.Mode()&(os.ModeNamedPipe|os.ModeSocket) != 0
}

// newConsole builds the interactive program, and the interface behind it.
//
// The streams are parameters rather than the process's own, so that the
// interface can run with its input disabled and its output redirected. That is
// what lets the interactive path be tested with no terminal attached: keys
// arrive as messages, Events arrive as messages, and what a person would have
// read lands in a buffer.
//
// It starts styled for a dark terminal, which is what an unknown terminal is
// assumed to be, and asks the real one as its first act. The answer arrives as
// a message rather than as a return value, because asking is a sequence
// written to the terminal and read back — and the program owns the terminal
// from the moment it starts. Asking beside it would mean two readers of one
// input, and the one that is not the program wins the keystrokes it should not
// have seen.
func newConsole(ctx context.Context, e *eva, in io.Reader, out io.Writer) (*tea.Program, *console, error) {
	input := textinput.New()
	input.Prompt = "› "
	input.Placeholder = "ask something"

	c := &console{
		ask:     e.answer,
		control: e,
		ctx:     ctx,
		input:   input,
		profile: colorprofile.Detect(out, os.Environ()),
		dark:    true,
		styles:  newStyles(true),
	}

	renderer, err := render.New(c, c.dark)
	if err != nil {
		return nil, nil, err
	}
	c.renderer = renderer

	// The program does not exist until the model does, and the reader has to
	// be able to end the program it is read by. The closure is what ties the
	// knot; it is called from the read loop, which the program starts.
	var program *tea.Program
	if file, isFile := in.(*os.File); isFile {
		quit := func() { program.Quit() }
		if pollable(file) {
			in = ended{File: file, quit: quit}
		} else {
			in = endedFile{Reader: file, quit: quit}
		}
	}
	program = tea.NewProgram(c, tea.WithContext(ctx), tea.WithInput(in), tea.WithOutput(out))

	// Two supplies, one doorway. The Recorder publishes to the feed and the
	// turn tells the second what is arriving; both reach the interface as a
	// message from outside the update loop, and the interface keeps them apart
	// because they are different claims.
	e.watch(
		&feed{program: program},
		func(text string) { program.Send(chunk{text: text}) },
	)

	return program, c, nil
}

// Init focuses the prompt, asks the terminal what colour it is, and says how
// to leave.
func (c *console) Init() tea.Cmd {
	return tea.Batch(
		c.input.Focus(),
		tea.RequestBackgroundColor,
		tea.Println(c.styles.hint.Render("eva — enter to send, /help for the commands, ctrl+c to interrupt a turn, ctrl+d to leave")),
	)
}

// Update folds one message into the interface.
func (c *console) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
	switch msg := msg.(type) {
	case tea.WindowSizeMsg:
		c.height = msg.Height
		if msg.Width > 0 {
			c.input.SetWidth(msg.Width)
		}

	case tea.BackgroundColorMsg:
		return c, c.background(msg.IsDark())

	case tea.KeyPressMsg:
		return c, c.key(msg)

	case chunk:
		c.arriving.WriteString(msg.text)
		return c, nil

	case committed:
		return c, c.fold(msg.event)

	case answered:
		return c, c.close(msg)
	}

	var cmd tea.Cmd
	c.input, cmd = c.input.Update(msg)
	return c, cmd
}

// View is the live area: what is arriving while a turn runs, and the prompt
// when none does.
//
// The turns already answered are not here. They are above the view, put there
// as each Run closed, which is what leaves them in the terminal's own
// scrollback rather than in a buffer this program has to hold and redraw.
func (c *console) View() tea.View {
	if !c.busy() {
		return tea.NewView(c.input.View())
	}

	var live strings.Builder
	if arriving := strings.TrimRight(c.arriving.String(), "\n"); arriving != "" {
		live.WriteString(c.tail(arriving))
		live.WriteString("\n")
	}
	live.WriteString(c.styles.hint.Render("answering — ctrl+c to interrupt"))
	return tea.NewView(live.String())
}

// liveLines is how many lines of an arriving answer the view shows when the
// window has not said how tall it is.
const liveLines = 10

// tail is the end of what has arrived, at most as many lines as there is room
// for.
//
// The live area is a window on a stream and not a transcript. Showing the
// whole of a long answer would grow the view past the window, and the whole of
// it is put above the view anyway once the Run closes — rendered, which is
// what a person actually reads it as.
func (c *console) tail(arriving string) string {
	room := liveLines
	if c.height > 0 && c.height-2 < room {
		room = c.height - 2
	}
	if room < 1 {
		room = 1
	}

	lines := strings.Split(arriving, "\n")
	if len(lines) <= room {
		return arriving
	}
	return strings.Join(lines[len(lines)-room:], "\n")
}

// busy reports whether a Run is in flight.
func (c *console) busy() bool { return c.cancel != nil }

// stop ends the Run in flight, if there is one.
//
// It is also what the caller uses once the program has ended, which is safe
// for the reason it would not otherwise be: the update loop runs on the
// goroutine that started the program, so by the time that call returns there
// is nothing else touching this.
func (c *console) stop() {
	if c.cancel != nil {
		c.cancel()
		c.cancel = nil
	}
}

// background takes the terminal's answer about its own colour, and restyles
// everything that follows from it. Nothing configures this: a style a person
// had to select is a style most people never select.
func (c *console) background(dark bool) tea.Cmd {
	if dark == c.dark {
		return nil
	}
	c.dark = dark
	c.styles = newStyles(dark)

	if dark {
		c.input.SetStyles(textinput.DefaultDarkStyles())
	} else {
		c.input.SetStyles(textinput.DefaultLightStyles())
	}

	if err := c.renderer.Background(dark); err != nil {
		c.blame(err)
	}
	return c.print()
}

// key acts on a key press.
func (c *console) key(k tea.KeyPressMsg) tea.Cmd {
	switch k.String() {
	case "ctrl+c":
		// During a turn this cancels it and gives the prompt back, rather
		// than ending the Session. A long wrong answer should cost seconds,
		// and the Session is a fold over what was committed — so the part of
		// the answer that did arrive is part of the transcript, and the next
		// turn is conditioned on it.
		if c.busy() {
			c.cancel()
			return nil
		}
		return tea.Quit

	case "ctrl+d":
		c.stop()
		return tea.Quit

	case "enter":
		if c.busy() {
			return nil
		}
		prompt := strings.TrimSpace(c.input.Value())
		if prompt == "" {
			return nil
		}
		c.input.Reset()
		c.put(c.styles.said.Render("› " + prompt))

		// A slash is a person addressing eva rather than the model, and what it
		// asks for happens here: no Run opens, nothing is committed, and the
		// Trace holds no record of a turn that never ran.
		if strings.HasPrefix(prompt, "/") {
			c.put(c.obey(prompt))
			return c.print()
		}
		return tea.Sequence(c.print(), c.start(prompt))
	}

	if c.busy() {
		return nil
	}

	var cmd tea.Cmd
	c.input, cmd = c.input.Update(k)
	return cmd
}

// start opens a Run for the prompt.
//
// The Run is counted before the goroutine is made rather than inside it: this
// happens in the update loop, which has ended by the time anything waits, so
// there is no Run the wait can miss.
func (c *console) start(prompt string) tea.Cmd {
	ctx, cancel := context.WithCancel(c.ctx)
	c.cancel = cancel
	c.arriving.Reset()
	c.running.Add(1)

	return func() tea.Msg {
		defer c.running.Done()
		defer cancel()

		outcome, err := c.ask(ctx, prompt)
		return answered{outcome: outcome, err: err}
	}
}

// fold takes one committed Event.
//
// The Renderer is fed here rather than subscribed to the Recorder directly,
// because what it produces has to go above this program's view and only the
// update loop may put it there.
// The live area is not fed from here. A Text record is a whole content block,
// which for an ordinary answer is the whole answer, and a person watching that
// arrive would watch nothing until it was over.
func (c *console) fold(e events.Event) tea.Cmd {
	if err := c.renderer.Committed(c.ctx, e); err != nil {
		c.blame(err)
	}
	return c.print()
}

// close ends the turn the interface was waiting on.
func (c *console) close(a answered) tea.Cmd {
	c.stop()
	c.arriving.Reset()

	// A turn that did not answer says why in its own claim, and that claim is
	// the one the Trace holds. The interface reads it rather than working the
	// reason out a second time from the error: "interrupted" for a turn
	// somebody stopped, the provider's own words for one that broke. A
	// classification made here could disagree with the record, and a person
	// reading the screen would have no way to know which of the two was true.
	//
	// A turn that broke leaves the Session standing either way. A console that
	// ended on the first provider failure would throw away a conversation over
	// something the next prompt might not hit.
	if a.outcome.Result != events.ResultDone && a.outcome.Summary != "" {
		c.put(c.styles.hint.Render(a.outcome.Summary))
	}

	// An error is not an Outcome. It is the record failing, which is the one
	// thing a person cannot be shown from the Trace.
	if a.err != nil {
		c.blame(a.err)
	}

	return tea.Batch(c.print(), c.input.Focus())
}

// put queues a block to go above the view, with its colour reduced to what the
// destination can show.
//
// The reduction happens here rather than on the way out. A program renders its
// own view against what the terminal can show, and puts what it is handed
// above that view exactly as it was handed it — so a turn nobody reduced would
// reach a sixteen-colour terminal in twenty-four-bit colour.
func (c *console) put(block string) {
	var adapted strings.Builder
	reduce := colorprofile.Writer{Forward: &adapted, Profile: c.profile}
	// Neither a builder nor the reducer in front of it has anywhere to fail.
	_, _ = io.WriteString(&reduce, strings.TrimRight(block, "\n"))

	c.pending = append(c.pending, adapted.String())
}

// blame queues a failure to go above the view, in the interface's own voice.
// A console that swallowed one would leave a person waiting at a prompt for a
// turn that is not coming.
func (c *console) blame(err error) {
	c.put(c.styles.hint.Render("eva: " + err.Error()))
}

// Show takes a finished turn from the Renderer. It is the Screen a Renderer
// built by newConsole writes to.
func (c *console) Show(turn string) error {
	c.put(turn)
	return nil
}

var _ render.Screen = (*console)(nil)

// print asks the program to put what is queued above its view.
//
// In sequence rather than in a batch: a batch runs its commands at once, and
// what a person reads is an order.
func (c *console) print() tea.Cmd {
	if len(c.pending) == 0 {
		return nil
	}

	blocks := c.pending
	c.pending = nil

	cmds := make([]tea.Cmd, len(blocks))
	for i, block := range blocks {
		cmds[i] = tea.Println(block)
	}
	return tea.Sequence(cmds...)
}

File: plugins/tui/src/console.ts
import {
  spendOf,
  type CostSummary,
  type Payload,
  type SessionID,
  type TranscriptMessage,
} from "@missingstudio/eva-schema"
import {
  costText,
  seconds,
  tokenLine,
  tookText,
  type Frame,
  type Overlay,
  type ThemeColors,
  type Work,
} from "@missingstudio/eva-tui-core"
import { refiltered, stepped as moved, type OpenOverlay } from "./overlay.js"

/**
 * The Console's state and the rules it changes by. Everything the screen
 * shows is a fold over these events, so every rule of the screen is a case
 * here — pure, and tested without a renderer, a Session API, or a clock.
 * The surface carries events in and frames out, and holds nothing of its
 * own.
 */
export interface ConsoleState {
  readonly session: SessionID
  // What the person has typed so far. The line editor changes it; a redraw
  // never loses it.
  readonly buffer: string
  // Where the caret sits in the buffer, counted in code points.
  readonly cursor: number
  /**
   * Every line submitted, oldest first, consecutive duplicates stored once.
   * `recall` is where a history walk stands — an index into `history`, and
   * absent when the person is not walking it. An edit to the buffer ends
   * the walk; moving the caret through a recalled line does not.
   */
  readonly history: readonly string[]
  readonly recall?: number
  // A question is open, and the next line answers it rather than opening a
  // Run.
  readonly asking: boolean
  /**
   * Esc has been pressed once against an open Run, so the next one
   * interrupts it. It decays on any other key: an interrupt is never
   * something a person arrives at by forgetting they armed it.
   */
  readonly armed: boolean
  // The one panel, when one is open. At most one, because a panel over a
  // panel is a screen to get lost in.
  readonly overlay?: OpenOverlay
  /**
   * Completion was dismissed for the line as it stands. The next edit
   * lifts it: dismissing is about this line, not about completion.
   */
  readonly hushed: boolean
  /**
   * The colors the screen is painted in. Absent keeps the renderer's own,
   * so a build with no theme Domain still draws. They are state rather than
   * something settled when a renderer started: a theme picker that cannot
   * repaint is a theme picker nobody can see.
   */
  readonly theme?: ThemeColors
  readonly live: string
  readonly model: string
  readonly cost: string
  readonly tokens: string
  readonly took: string
  readonly mode: string
  // When the open Run opened. Read only while `work.running`.
  readonly opened: number
  readonly work: Work
  // The record's fold, and nothing else. A fold replaces it, so nothing
  // here is the source of truth for long.
  readonly shown: readonly TranscriptMessage[]
  /**
   * What this Console said of its own — command output, a notice, a
   * question. They never enter `shown`, because a line the record cannot
   * rebuild is not the record. They last until the conversation moves on: a
   * Run opens, or the Console follows another Session.
   */
  readonly notes: readonly string[]
}

/** Where this Console runs, which no Run changes. */
export interface Place {
  readonly version: string
  readonly branch: string
  readonly directory: string
}

export type ConsoleEvent =
  // The line editor changed the buffer, or moved the caret through it.
  | { readonly kind: "typed"; readonly buffer: string; readonly cursor: number }
  // The person walked the prompt history.
  | { readonly kind: "recalled"; readonly direction: "back" | "forward" }
  // A line left the editor for the loop, whatever it turns out to mean.
  | { readonly kind: "submitted"; readonly line: string }
  // The surface says something of its own: command output, a notice.
  | { readonly kind: "said"; readonly text: string }
  // Eva asked a question; the next line answers it.
  | { readonly kind: "asked"; readonly question: string }
  | { readonly kind: "answered" }
  // A Run opened on this prompt.
  | { readonly kind: "opened"; readonly line: string; readonly at: number }
  // The open Run streamed a payload into the Live area.
  | { readonly kind: "streamed"; readonly payload: Payload }
  // The clock turned while the Run was open.
  | { readonly kind: "ticked"; readonly at: number }
  // The open Run closed; the fold that replaces its stream follows.
  | { readonly kind: "closed"; readonly at: number }
  // A panel opened over the fold, and what taking a row from it means.
  | { readonly kind: "opened-overlay"; readonly overlay: OpenOverlay }
  // The panel's own query changed. A `buffer` panel is refiltered by the
  // line instead, on `typed`.
  | { readonly kind: "filtered"; readonly query: string }
  // The selection moved through the panel's rows.
  | { readonly kind: "stepped"; readonly by: number }
  // The panel closed, whether a row was taken or not.
  | { readonly kind: "closed-overlay" }
  // The person pressed the key that steps back. What that means is the
  // stack's to decide, and `backStep` decides it.
  | { readonly kind: "backed" }
  // A key that was not the step-back key arrived, so the interrupt that was
  // armed is not armed any more.
  | { readonly kind: "disarmed" }
  // The screen's colors changed: a theme was chosen, or one is being looked
  // at. No colors is the renderer's own default, which is what restoring
  // says when nothing was set before.
  | { readonly kind: "themed"; readonly colors?: ThemeColors }
  // The person cancelled whatever was open.
  | { readonly kind: "cancelled" }
  // The record's fold replaces what was shown.
  | {
      readonly kind: "folded"
      readonly messages: readonly TranscriptMessage[]
      readonly model: string
      readonly summary: CostSummary
      readonly holding: boolean
    }
  // A command opened another Session; the Console follows it there.
  | { readonly kind: "selected"; readonly session: SessionID }

const IDLE: Work = { running: false, elapsed: "", tick: 0, hint: "" }

export const initial = (session: SessionID): ConsoleState => ({
  session,
  buffer: "",
  cursor: 0,
  history: [],
  asking: false,
  armed: false,
  hushed: false,
  live: "",
  model: "",
  cost: "",
  tokens: "",
  took: "",
  mode: "ready",
  opened: 0,
  work: IDLE,
  shown: [],
  notes: [],
})

const said = (author: TranscriptMessage["author"], text: string): TranscriptMessage => ({
  author,
  blocks: [{ type: "content", block: 0, content: { type: "text", text } }],
})

/**
 * What the Console says of its own, one note per line. A blank line inside
 * one saying says nothing, so it is left out rather than drawn empty — but
 * one saying stands clear of the one before it, because two answers with
 * nothing between them read as one longer answer.
 */
const say = (state: ConsoleState, text: string): ConsoleState => {
  const lines = text.split("\n").filter((line) => line !== "")
  if (lines.length === 0) return state
  const apart = state.notes.length === 0 ? [] : [""]
  return { ...state, notes: [...state.notes, ...apart, ...lines] }
}

/**
 * What one press of the step-back key does, from what is open. One key, one
 * meaning: step back. The reducer applies it and the surface reads it to
 * know whether to interrupt — one rule, in one place, read twice, because a
 * screen that says one thing and a Run that does another is the whole
 * failure this stack exists to avoid.
 */
export type BackStep = "close-overlay" | "clear-line" | "arm" | "interrupt" | "nothing"

export const backStep = (state: ConsoleState): BackStep => {
  if (state.overlay !== undefined) return "close-overlay"
  if (state.buffer !== "") return "clear-line"
  if (!state.work.running) return "nothing"
  return state.armed ? "interrupt" : "arm"
}

// What the status line says an armed interrupt is waiting for. The surface
// chooses the words, so every renderer says the same ones.
export const ARMED = "esc again to interrupt"

// What the status line says while Eva is waiting on an answer. The left half
// holds one state at a time, and this is the one that outranks `ready`: the
// line is open, but not for a prompt.
export const ASKING = "answer the question above"

// A walk that has ended leaves no index behind.
const settled = (state: ConsoleState): ConsoleState => {
  const { recall: _recall, ...rest } = state
  return rest
}

const disarm = (state: ConsoleState): ConsoleState =>
  state.armed ? { ...state, armed: false, work: { ...state.work, hint: "" } } : state

// A panel that has closed leaves nothing behind.
const shut = (state: ConsoleState): ConsoleState => {
  const { overlay: _overlay, ...rest } = state
  return rest
}

// One step back. The interrupt itself is the surface's to perform — what it
// leaves on the screen is the cancel that follows it.
const back = (state: ConsoleState): ConsoleState => {
  switch (backStep(state)) {
    case "close-overlay":
      // Dismissing completion is about this line: the next edit asks for
      // it again, and until then the line is the person's alone.
      return { ...shut(state), hushed: state.overlay?.source === "buffer" }
    case "clear-line":
      return { ...settled(state), buffer: "", cursor: 0, hushed: false }
    case "arm":
      return { ...state, armed: true, work: { ...state.work, hint: ARMED } }
    case "interrupt":
    case "nothing":
      return state
  }
}

const walked = (state: ConsoleState, direction: "back" | "forward"): ConsoleState => {
  if (direction === "back") {
    // A walk starts behind the newest line and stops at the oldest — going
    // past it would wrap, and a wrap reads as the history changing.
    const index = (state.recall ?? state.history.length) - 1
    const line = state.history[index]
    if (line === undefined) return state
    return { ...state, buffer: line, cursor: Array.from(line).length, recall: index }
  }
  if (state.recall === undefined) return state
  const line = state.history[state.recall + 1]
  // Forward past the newest line is the empty line the walk started from.
  if (line === undefined) return { ...settled(state), buffer: "", cursor: 0 }
  return { ...state, buffer: line, cursor: Array.from(line).length, recall: state.recall + 1 }
}

export const apply = (state: ConsoleState, event: ConsoleEvent): ConsoleState => {
  switch (event.kind) {
    case "typed": {
      // Moving the caret is not editing: it ends no history walk, asks for
      // no completion, and leaves a dismissal in place.
      if (event.buffer === state.buffer) {
        return { ...state, cursor: event.cursor }
      }
      // A panel that follows the line follows it here. The surface decides
      // whether the line still asks for one; this only keeps the open one
      // in step with what it is completing.
      const typed = { ...settled(state), buffer: event.buffer, cursor: event.cursor, hushed: false }
      return typed.overlay === undefined || typed.overlay.source !== "buffer"
        ? typed
        : { ...typed, overlay: refiltered(typed.overlay, event.buffer) }
    }
    case "recalled":
      return walked(state, event.direction)
    case "opened-overlay":
      return { ...state, overlay: event.overlay }
    case "filtered":
      return state.overlay === undefined
        ? state
        : { ...state, overlay: refiltered(state.overlay, event.query) }
    case "stepped":
      return state.overlay === undefined
        ? state
        : { ...state, overlay: moved(state.overlay, event.by) }
    case "closed-overlay":
      return shut(state)
    case "backed":
      return back(state)
    case "disarmed":
      return disarm(state)
    case "themed": {
      const { theme: _theme, ...rest } = state
      return event.colors === undefined ? rest : { ...rest, theme: event.colors }
    }
    case "submitted": {
      const history =
        state.history[state.history.length - 1] === event.line
          ? state.history
          : [...state.history, event.line]
      return { ...settled(state), history }
    }
    case "said":
      return say(state, event.text)
    case "asked":
      return { ...say(state, event.question), asking: true }
    case "answered":
      return { ...state, asking: false }
    case "opened":
      return {
        // A panel over a moving fold is a panel over the wrong thing, so
        // the Run that starts takes it with it.
        ...shut(state),
        shown: [...state.shown, said("human", event.line)],
        // The conversation moves on, so what the Console said before it
        // goes with it.
        notes: [],
        mode: "running",
        live: "",
        took: "",
        opened: event.at,
        hushed: false,
        work: { running: true, elapsed: seconds(0), tick: 0, hint: "" },
      }
    case "streamed":
      // The stream lands in the Live area, never in the fold. Anything in
      // the session pane came from the record.
      return event.payload.kind === "text" && event.payload.content.type === "text"
        ? { ...state, live: state.live + event.payload.content.text }
        : state
    case "ticked":
      // A tick after the Run closed says nothing. The spinner and the
      // caption both read off the count, so both renderers animate in step.
      return state.work.running
        ? {
            ...state,
            work: {
              running: true,
              elapsed: seconds(event.at - state.opened),
              tick: state.work.tick + 1,
              hint: state.work.hint,
            },
          }
        : state
    case "closed":
      return { ...state, work: IDLE, took: tookText(event.at - state.opened) }
    case "cancelled":
      // The Run the interrupt was armed against is gone, so the arming
      // goes with it rather than waiting for the next one.
      return { ...state, mode: "ready", live: "", work: IDLE, armed: false }
    case "folded": {
      /**
       * The fold, drawn over what the Console was showing. A Run that just
       * closed is in the record, so a fold that comes back with nothing has
       * lost the record rather than the conversation: `holding` keeps the
       * screen instead of blanking it. Choosing another Session passes it
       * over, because an empty Session really is empty.
       */
      const shown = event.messages.length > 0 || !event.holding ? event.messages : state.shown
      return {
        ...state,
        shown,
        model: event.model,
        cost: costText(spendOf(event.summary, shown.length > 0)),
        tokens: tokenLine(event.summary.inputTokens, event.summary.outputTokens),
        live: "",
        // A fold while a Run is open is a repaint, not an ending: the pipe
        // dropped and the record took the stream's place. The Run says when
        // it is over, and it has not.
        mode: state.work.running ? state.mode : "ready",
      }
    }
    case "selected":
      // Another Session is another conversation, so the words this one said
      // do not follow it there.
      return { ...state, session: event.session, notes: [] }
  }
}

// What a renderer is given of the panel: what is drawn, and nothing that
// is decided. Where a chosen row goes is the surface's business.
const drawable = ({ intent: _intent, all: _all, ...overlay }: OpenOverlay): Overlay => overlay

export const frameOf = (state: ConsoleState, place: Place): Frame => ({
  banner: { ...place, model: state.model },
  session: state.shown,
  notes: state.notes,
  live: state.live,
  input: state.buffer,
  cursor: state.cursor,
  took: state.took,
  work: state.work,
  status: {
    model: state.model,
    tokens: state.tokens,
    cost: state.cost,
    // One state at a time on the left half, and a question outranks the
    // rest: what the line is for right now is what it says.
    mode: state.asking ? ASKING : state.mode,
  },
  ...(state.overlay === undefined ? {} : { overlay: drawable(state.overlay) }),
  ...(state.theme === undefined ? {} : { theme: state.theme }),
})
